File: DAA/max-subarray-sum.cpp
#include <bits/stdc++.h>
using namespace std;

struct Item {
	// object to store sum value and left index and right index 
	int sum;
	int left;
	int right;
	Item(int sum, int left, int right) {
		this->sum = sum;
		this->left = left;
		this->right = right;
	}
};

Item maxCrossingSubarraySum(vector<int>& nums, int start, int mid, int end) {
	int leftSum = INT_MIN, rightSum = INT_MIN;
	int sum = 0;
	int maxLeft, maxRight;

	// finding left sum
	for(int i=mid; i>=0; i--) {
		sum += nums[i];
		if (sum > leftSum) {
			leftSum = sum;
			maxLeft = i;
		}
	}
	// finding right sum
	sum = 0;
	for(int i=mid+1; i<=end; i++) {
		sum += nums[i];
		if (sum > rightSum) {
			rightSum = sum;
			maxRight = i;
		}
	}
	return Item(leftSum+rightSum, maxLeft, maxRight);
}
Item maxSubarraySum(vector<int>& nums, int start, int end) {
	if (start == end) {
		// if there is only one item, return itself
		return Item(nums[start], start, end);
	}
	int mid = start + (end-start)/2;
	// find left sum
	Item left = maxSubarraySum(nums, start, mid);
	// find right sum
	Item right = maxSubarraySum(nums, mid+1, end);
	// find crossing sum
	Item cross = maxCrossingSubarraySum(nums, start, mid, end);

	// if left sum if greater than both right sum and cross sum, return it
	if (left.sum > right.sum && left.sum > cross.sum)
		return left;
	// if right sum if greater than both left sum and cross sum, return it
	else if (right.sum > left.sum && right.sum > cross.sum)
		return right;
	// else return the cross sum
	else
		return cross;
}

int main() {
	
	int n; cin >> n;
	vector<int> nums(n);
	for(int i=0; i<n; i++)
		cin >> nums[i];

	Item result = maxSubarraySum(nums, 0, n-1);
	cout << result.sum << "\n";
	for(int i=result.left; i<=result.right; i++)
		cout << nums[i] << " ";

	cout << endl;


	return 0;
}

File: DAA/PYTHON/max-subarray-sum.py

class Pair:
  def __init__(self, summ, first, second):
    self.summ = summ
    self.first = first 
    self.second = second

def maxSubarrayCrossingSum(arr, start, mid, end):
  leftSum = float('-inf')
  summ = 0
  maxLeft, maxRight = 0, 0
  for i in range(mid, -1, -1):
    summ += arr[i]
    if summ > leftSum:
      leftSum = summ
      maxLeft = i
  summ = 0
  rightSum = float('-inf')
  for j in range(mid + 1, end + 1):
    summ += arr[j]
    if summ > rightSum:
      rightSum = summ
      maxRight = j
  
  return Pair(leftSum + rightSum, maxLeft, maxRight)

def maxSubarraySum(arr, start, end):
  if (start == end):
    return Pair(arr[start], start, end) 
  else:
    mid = start + (end-start) // 2

    left = maxSubarraySum(arr, start, mid)
    right = maxSubarraySum(arr, mid + 1, end)
    cross = maxSubarrayCrossingSum(arr, start, mid, end)

    if left.summ > right.summ and left.summ > cross.summ:
      return left
    elif right.summ > left.summ and right.summ > cross.summ:
      return right
    else:
      return cross


arr = [1,2,3,-4]
start = 0
end = len(arr) - 1

result = maxSubarraySum(arr, start, end)

print(result.summ)
for i in range(result.first, result.second + 1):
  print(arr[i], end = " ")
print()






